tools/codegen/dest/lazy_ir.py

```diff
@@ -39,10 +54,41 @@ def node_ctor_inputs(func: LazyIrSchema) -> str:
     """
     Produce a formatted string with the arguments as passed into the constructor of a node class.
     """
-    node_ctor_values = [node_ctor_arg_rvalue_string(arg) for arg in func.filtered_types()]
+    node_ctor_values = [node_ctor_arg_rvalue_string(arg) for arg in func.filtered_types()]+ ["node_hash", "dag_hash"]
     return ",\n                              ".join(node_ctor_values)
 
 
+def cached_shape_inference(func: NativeFunction, schema: LazyIrSchema,
+                           returns_length: int, all_types: List[NamedCType]) -> str:
+    # call the meta kernel if it exists, to compute output shape/dtype for our IR
+    if func.structured or func.structured_delegate is not None:
+        meta_out = """cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>(std::initializer_list<Shape>{Shape(out_meta.scalar_type(), out_meta.sizes().vec())}));"""
+        if returns_length > 1:
+            def this_shape(i: int) -> str:
+                return f"Shape(std::get<{i}>(out_meta).scalar_type(), std::get<{i}>(out_meta).sizes().vec())"
+            inner_shapes_str = ','.join([this_shape(i) for i in range(returns_length)])
+            meta_out = "cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>(std::initializer_list<Shape>{" + inner_shapes_str + "}));"
+
+        shapes_body = f"""
+            auto out_meta = at::meta::{schema.aten_name}({', '.join(str(t.name) for t in all_types)});
+            {meta_out}
+"""
+    else:
+        shape_sig = ComputeShapeSignature(func)
+        shapes_body = f"""
+            cached_shapes = shape_cache->Add(dag_hash, std::make_shared<std::vector<Shape>>({shape_sig.shape_call}));
+"""    
+    
+    shapes_str = f"""
+        auto shape_cache = torch::lazy::GetShapeCache();
+        auto cached_shapes = shape_cache->Get(dag_hash);
+        if (cached_shapes == nullptr) {{{shapes_body}
```

Comment:
Nit: the formatting is a little strange here.

Reply:
this was at least intentional, 

shapes_body starts out with a newline and then always has the same number of indents

leaving the }} to the next line makes it show up correctly.